types.d.ts

```diff
@@ -0,0 +1,699 @@
+import { EventEmitter } from "events";
+import https from 'https';
+import * as core from 'express-serve-static-core'
+
+//TODO: need to better understand this type
+type InputType = {}
+
+
+export interface Subscriber {
+  send(message: string): void
+}
+
+export interface ActionTypeClass {
+  new(thing: Thing, input: InputType): ActionTypeClass
+}
+
+/**
+ * An Action represents an individual action on a thing.
+ */
+export declare class Action {
+
+  constructor(id: string, thing: Thing, name: string, input: InputType);
+
+  /**
+   * Get the action description.
+   *
+   * @returns {Object} Description of the action as an object.
+   */
+  asActionDescription(): Action.ActionDescription;
+
+  /**
+   * Set the prefix of any hrefs associated with this action.
+   *
+   * @param {String} prefix The prefix
+   */
+  setHrefPrefix(prefix: string);
+
+  /**
+   * Get this action's ID.
+   *
+   * @returns {String} The ID.
+   */
+  getId(): string
+
+  /**
+   * Get this action's name.
+   *
+   * @returns {String} The name.
+   */
+  getName(): string
+
+  /**
+   * Get this action's href.
+   *
+   * @returns {String} The href.
+   */
+  getHref(): string
+
+  /**
+   * Get this action's status.
+   *
+   * @returns {String} The status.
+   */
+  getStatus(): string;
+
+  /**
+   * Get the thing associated with this action.
+   *
+   * @returns {Object} The thing.
+   */
+  getThing(): Thing
+
+  /**
+   * Get the time the action was requested.
+   *
+   * @returns {String} The time.
+   */
+  getTimeRequested(): string
+  /**
+   * Get the time the action was completed.
+   *
+   * @returns {String} The time.
+   */
+  getTimeCompleted(): string
+  /**
+   * Get the inputs for this action.
+   *
+   * @returns {Object} The inputs.
+   */
+  getInput(): InputType
+
+  /**
+   * Start performing the action.
+   */
+  start(): void;
+
+  /**
+   * Override this with the code necessary to perform the action.
+   *
+   * @returns {Object} Promise that resolves when the action is finished.
+   */
+  performAction(): Promise<any>;
+
+  /**
+   * Override this with the code necessary to cancel the action.
+   *
+   * @returns {Object} Promise that resolves when the action is cancelled.
+   */
+  cancel(): Promise<any>;
+
+  /**
+   * Finish performing the action.
+   */
+  finish(): void
+}
+
+declare namespace Action {
+  //TODO need to better understand this type
+  type ActionMetadata = any
+
+  //TODO need to better understand this type
+  interface ActionDescription {
+    [name: string]: {
```

Comment:
This feels like a byproduct of inconsistency within the API where ActionDescription has the name key included in its getter and PropertyDescription doesn't. I think it might be more clear for ActionDescription to not include the key and put it on the getter instead. This would potentially make the Thing's properties, actions, and events members be consistently typed as maps from name to description or metadata

Reply:
I'm good either way! Is there an action I should take on this?

Reply:
@hobinjk I just realized I didn't follow up on this. Is there some type of action you'd like from me in the typings to address this?

Reply:
Oh, sorry for not responding earlier! Looking at this again I think the current type is good enough